### PR TITLE
[IMP] core: avoid unnecessary recomputes on load

### PR DESCRIPTION
**Pull request by aj-fuentes**
When loading records from xml if we blindly write the values to the DB
we may trigger expensive recomputations for values that didn't change.
Writing without checks has the advantage of not fetching extra data from
the DB, thus it's expected to be faster. That premise breaks as soon as
there are triggers for computed stored fields. The triggers can be much
more expensive since they require many reads and can potentially involve
millions of records.

Here we implement a simple scheme to avoid writes only when we are
*sure* nothing is changing. There are many pathological cases with
computed fields that we wish to avoid thus we restrict to ensuring that
in the current write nothing changes. If we cannot ensure it we let it
happen. Note that the fact that we *cannot* ensure that nothing changes
doesn't mean that it *will be* a change. The check aims to be fast to
avoid slowing down the load process too much.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/models.py [hunk 1]
```diff
@@ -5841,6 +5892,7 @@ def __eq__(self, other):
         if not isinstance(other, BaseModel):
             if other:
                 filename, lineno = frame_codeinfo(currentframe(), 1)
+                import pdb;pdb.set_trace()
```
Comment (KangOl) [inline odoo/models.py:5895]: Will block runbot :p

Comment (aj-fuentes) [reply odoo/models.py:5895]: 🤦🏼

**Review by rco-odoo — COMMENTED** [review]
Overall it looks good.  My remarks below are only about code style and improvement.

File: addons/mrp/data/mrp_data.xml [hunk 2]
```diff
@@ -14,9 +14,10 @@
         </record>
 
         <!-- Enable the manufacturing in warehouse0 -->
-        <record id='stock.warehouse0' model='stock.warehouse'>
-            <field name='manufacture_to_resupply' eval='True'/>
-        </record>
+        <function model="stock.warehouse" name="write">
+            <value eval="[ref('stock.warehouse0')]"/>
+            <value name="vals" eval="{'manufacture_to_resupply': True}"/>
+        </function>
```
Comment (rco-odoo) [inline addons/mrp/data/mrp_data.xml:20]: Obscurity++

File: odoo/models.py [hunk 3]
```diff
@@ -4415,9 +4415,63 @@ def _load_records(self, data_list, update=False):
                 imd.browse(d_id).unlink()
                 to_create.append(data)
 
+        def must_update(fname, val, rec):
+            if fname not in rec._fields:
+                return True
+
+            field = rec._fields[fname]
+            if val is None or val is False:
+                return rec[fname] is not False
```
Comment (rco-odoo) [inline odoo/models.py:4424]: Not correct for relational fields.

File: odoo/models.py [hunk 4]
```diff
@@ -4415,9 +4415,63 @@ def _load_records(self, data_list, update=False):
...
+
+                if len(val) > 1:
+                    # multiple commands are complex to analyze, this optimization aims to
+                    # analyze only single-command updates which is fine since that's the
+                    # most common case in xml data
+                    return True
+
+                command, = val
+                # we want the linked records even if unactive
+                rec_val = rec.with_context(active_test=False)[field.name]
```
Comment (rco-odoo) [inline odoo/models.py:4441]: Why do you compare the field's value from another context?  This looks inconsistent to me.

File: odoo/models.py [hunk 5]
```diff
@@ -4415,9 +4415,63 @@ def _load_records(self, data_list, update=False):
...
+                    # link something already linked
+                    return False
+                elif command[0] == Command.UNLINK and command[1] not in rec_val.ids:
+                    # unlink something that wasn't linked
+                    return False
+
+                # for any other command/case we are unsure there is no change
+                return True
+            elif field.type == 'many2one':
+                return rec.with_context(active_test=False)[fname].ids != [val]
```
Comment (rco-odoo) [inline odoo/models.py:4455]: Nonsense, as `active_test` has no effect on `many2one` fields.

**Review by rco-odoo — COMMENTED** [review]
Let me push a simplification (and correction) of the code.

File: odoo/models.py (hunk 4 above)
Comment (rco-odoo) [reply odoo/models.py]: It is actually inconsistent for `Command.CLEAR`, but not for `Command.LINK` and `Command.UNLINK`.

File: addons/mrp/data/mrp_data.xml (hunk 2 above)
Comment (aj-fuentes) [reply addons/mrp/data/mrp_data.xml:20]: At least it's clear that we are forcing the write :) ...

**Comment by aj-fuentes** [comment]
@rco-odoo would you check the last commit I pushed. It's simpler... **if** correct :/
We may still limit to len=1 changes if you believe is safer.

File: odoo/models.py [hunk 6]
```diff
@@ -4415,9 +4415,28 @@ def _load_records(self, data_list, update=False):
...
+            """ Determine whether ``rec[fname]`` must be updated with ``val``. """
+            field = rec._fields.get(fname)
+            if field is None:
+                return True
+
+            if val is None or val is False:
+                return not rec[fname]
+
+            if field.type in ('many2many', 'one2many'):
+                return field.convert_to_cache([], rec) != field.convert_to_cache(val, rec)
```
Comment (rco-odoo) [inline odoo/models.py:4428]: By the way, the left-hand expression is 100% equivalent to `field.convert_to_cache(rec[fname], rec)`.  Which means that both the `if` and `else` clauses can be identical.  And I believe the `if val is None...` above is useless too, as most methods `field.convert_to_cache()` contain similar code.

File: addons/mrp/data/mrp_data.xml (hunk 2 above)
Comment (aj-fuentes) [reply addons/mrp/data/mrp_data.xml:20]: I was referring to the facts that:
1. The write of that model is full of side effects, to records that are **not** updated by the write (basically all warehouses).
2. Setting the value is unneeded since many years (it's the default, the record entry is inside a noupdate block, thus it only kicks in at install, but it's there just for the side effects ??? :/).

With an explicit call to write (I agree in an obfuscated way) I hope people will at least take the signal to check the blame of the file to understand what's going on. Explicit comment can be added to the file.

In summary: before - inconspicuous side effects, after - it pops to the eye that something weird is going on.

Comment (rco-odoo) [reply addons/mrp/data/mrp_data.xml:20]: I personally prefer a comment that explains why it is written that way.  The comment may be concise, and more detailed information should be found in a commit message by blaming the file.  I find this more efficient , as it immediately reminds you the reason for the code to be that way.  This was the actual motivation behind my remark :wink:

**Comment by rco-odoo** [comment]
@aj-fuentes the last version **is** correct, almost by construction.  The question of the limit is more about whether it is _efficient_ to check it that way.  And for the latter, the way to know is benchmarking :stuck_out_tongue_winking_eye:

File: addons/mrp/data/mrp_data.xml (hunk 2 above)
Comment (KangOl) [reply addons/mrp/data/mrp_data.xml:20]: Another solution is to extract this "side effect" part into a private method and call it explicitly from the XML file.

Comment (aj-fuentes) [reply addons/mrp/data/mrp_data.xml:28]: > Another solution is to extract this "side effect" part into a private method and call it explicitly from the XML file.

It's complex. The side effects parts are called interleaved with other code. I experimented extracting some parts but TBH given that MRP is involved I feel it's too complex to change without proper functional knowledge. Basically, the write method is also a _fixer_ method that ensures warehouses (and other details) are properly configured. 

Some example side effects below (non-exhaustive). They depend on how the data is in the DB. Which makes it worse: if we install stock first, change something, then install mrp.

[Side effect 1](https://github.com/odoo/odoo/blob/a3da4e6beecbb4c1104269f7087eed01d95245f5/addons/mrp/models/stock_warehouse.py#L278-L280) is trivial, and there is already an extracted method. Note the common theme: the call to `_update_location_manufacture` is done unconditionally (even if the value is the same). 

[Side effect 2](https://github.com/odoo/odoo/blob/a3da4e6beecbb4c1104269f7087eed01d95245f5/addons/stock/models/stock_warehouse.py#L155-L156) note this is for all warehouses including inactive ones. It doesn't matter what field we write.

[Side effect 3](https://github.com/odoo/odoo/blob/a3da4e6beecbb4c1104269f7087eed01d95245f5/addons/stock/models/stock_warehouse.py#L184-L247) this is the trickiest because it triggers stuff based on dependencies. The field we write here matters.

The rest of the write method has more side effects. There may be more lurking around. 

We should take the warning that this change is prone to break models that have side effects on the write. The only one that breaks in our tests is stock.warehouse. We test with all modules installed, different combinations could lead to unexpected extra side effects.

Comment (aj-fuentes) [reply addons/mrp/data/mrp_data.xml:28]: > I personally prefer a comment that explains why it is written that way.

I added a comment.

**Comment by aj-fuentes** [comment]
@rco-odoo Thanks for all the hints and re-reviews :)

I will run a benchmark to see what we get. I'll test this also in some of the DBs that fail the upgrade due to big recomputes.
